Custom Summary Page for NotebookWizard and notebook Cell with wizard variables

ranasaria opened the pull request:
<!-- Thank you for submitting a Pull Request. Please:
* Read our Pull Request guidelines:
  https://github.com/Microsoft/azuredatastudio/wiki/How-to-Contribute#pull-requests.
* Associate an issue with the Pull Request.
* Ensure that the code is up-to-date with the `master` branch.
* Include a description of the proposed changes and how to test them.
-->

This PR contains 2 main things:
1. Changes needed to support custom various UI elements of Summary Page definition for NotebookWizard.
2. Adding wizard variable values in code cell in the notebook as python code.

Testing has been done by running the Arc control create and bdc notebook generation workflows.

Charles-Gagnon commented on extensions/resource-deployment/src/interfaces.ts:
```diff
@@ -109,10 +109,11 @@ export interface NotebookWizardInfo extends WizardInfoBase {
 	taskName?: string;
 	type?: DeploymentType;
 	runNotebook?: boolean;
+	codeCellInssertionPosition?: number;
```
codeCellInsertionPosition

Charles-Gagnon commented on extensions/resource-deployment/src/services/notebookService.ts:
```diff
@@ -54,6 +55,25 @@ export class NotebookService implements INotebookService {
+	public async insertCellAndLaunchNotebook(notebook: string, cellStatements: string[], insertionPosition: number = 0): Promise<void> {
+		const notebook_1 = await this.launchNotebook(notebook);
+		notebook_1.edit((editBuilder: azdata.nb.NotebookEditorEdit) => {
+			// 5 is the position after the 'Set variables' cell in the deployment notebooks
```
This comment seems like it belongs in deployClusterWizard instead

ranasaria replied:
good catching - fixing

Charles-Gagnon commented on extensions/resource-deployment/src/ui/deployClusterWizard/deployClusterWizard.ts:
```diff
@@ -137,16 +138,10 @@ export class DeployClusterWizard extends WizardBase<DeployClusterWizard, DeployC
-			});
-		}, (error) => {
-			vscode.window.showErrorMessage(error);
+		const variableeValueStatements = this.model.getCodeCellContentForNotebook();
```
variableValueStatements

ranasaria replied:
fixed

Charles-Gagnon commented on extensions/resource-deployment/src/ui/model.ts:
```diff
@@ -32,6 +33,24 @@ export class Model {
+				return `${varName} = '${value}'`;
+			});
+		statements.push(`print('Variables have been set successfully.')`);
+		return statements.map(line => line + EOL);
```
Does the EOL not get inserted automatically when you pass in an array of strings? Seems like that would be the behavior I expect.

If you do need to add them here just append the EOL when you create the strings above (lines 43 and 45)

ranasaria replied:
fixed

Charles-Gagnon commented on extensions/resource-deployment/src/ui/model.ts:
```diff
@@ -32,6 +33,24 @@ export class Model {
+			.filter(propertyName => propertyName.startsWith(NoteBookEnvironmentVariablePrefix))
+			.map(propertyName => {
+				const value = this.escapeForNotebookCodeCell(this.getStringValue(propertyName, ''));
+				const regex = new RegExp(`^${NoteBookEnvironmentVariablePrefix}`);
```
This regex can just be a const declared outside the class - no reason to make a separate one for every line.

Charles-Gagnon commented on extensions/resource-deployment/src/ui/model.ts:
```diff
@@ -41,4 +60,15 @@ export class Model {
 		});
 	}
+
+	public interpolateVariableValues(inputValue: string): string {
```
These functions could really use doc comments explaining what they're doing. And can you come up with a better name? "interpolate" is way too generic

Charles-Gagnon commented on extensions/resource-deployment/src/ui/modelViewUtils.ts:
```diff
@@ -106,12 +114,20 @@ export function createTextInput(view: azdata.ModelView, inputInfo: { defaultValu
+	}
+
+	if (info.color !== undefined) {
+		cssStyles['color'] = info.color;
```
If you're going to be passing in cssStyles as a general param anyways might as well just get rid of the specific style params and just combine them into the cssStyles param (fontStyle, fontWeight and color)

ranasaria replied:
I considered that. The idea is to have separate params for commonly used ones and use the cssStyles for the more esoteric ones. This makes for easier usage.

Charles-Gagnon replied:
At the cost of many more parameters and inconsistent handling of how styles are passed in. 

If you really want to make "common" styles more easily accessible you make make the parameter type have a set of predefined properties in addition to taking in extras such as 

`export type MyCssStyles {
    color?: string;
    fontWeight?: string;
    [key: string]: string;
}`

I'm fine with either - but both are better than having a mix of specified params and an object containing more.

ranasaria replied:
I like your suggestion of extending CSSStyles type to include named as well as generic properties to make this simpler and consistent. Will do the implementation accordingly.

Charles-Gagnon commented on extensions/resource-deployment/src/ui/modelViewUtils.ts:
```diff
@@ -215,16 +237,18 @@ export function initializeWizardPage(context: WizardPageContext): void {
+	context.sectionInfo.rowAlignItems = context.sectionInfo.rowAlignItems || DefaultRowAlignItems;
+	context.sectionInfo.rowWidth = context.sectionInfo.rowWidth || DefaultRowWidth;
+	context.sectionInfo.rowHeight = context.sectionInfo.rowHeight || DefaultRowHeight;
+	context.sectionInfo.inputWidth = context.sectionInfo.inputWidth || DefaultInputWidth;
```
You're duplicating the inputWidth and labelWidth here

ranasaria replied:
good catch will remove the duplication.

ranasaria replied:
fixed

Charles-Gagnon commented on extensions/resource-deployment/src/ui/modelViewUtils.ts:
```diff
@@ -256,24 +295,41 @@ function processFields(fieldInfoArray: FieldInfo[], components: azdata.Component
+	}
+	let flexComponent = view.modelBuilder.flexContainer().withItems(items, itemsStyle).withLayout(flexLayout).component();
+	if ( cssStyles ) {
+		flexComponent.CSSStyles = Object.assign(flexComponent.CSSStyles || {}, cssStyles);
```
It'd be better just to do this before creating the component and assigning it during creation

ranasaria replied:
will do.

ranasaria replied:
fixed

Charles-Gagnon commented on extensions/resource-deployment/src/ui/modelViewUtils.ts:
```diff
@@ -256,24 +295,41 @@ function processFields(fieldInfoArray: FieldInfo[], components: azdata.Component
 
 export function createGroupContainer(view: azdata.ModelView, items: azdata.Component[], layout: azdata.GroupLayout): azdata.GroupContainer {
-	return view.modelBuilder.groupContainer().withItems(items).withLayout(layout).component();
+	return view.modelBuilder.groupContainer().withItems(items).	withLayout(layout).component();
```
extra spaces here

ranasaria replied:
removed

Charles-Gagnon commented on extensions/resource-deployment/src/ui/modelViewUtils.ts:
```diff
@@ -256,24 +295,41 @@ function processFields(fieldInfoArray: FieldInfo[], components: azdata.Component
 	}
-	if (labelPosition && labelPosition === LabelPosition.Left) {
-		const row = createFlexContainer(view, inputs);
+	if (inputs.length > 0 && fieldInfo.labelPosition && fieldInfo.labelPosition === LabelPosition.Left) {
```
Inputs.length is always going to be > 0 since we start off with the label component in there

And you can just shorten this to fieldInfo.labelPosition === LabelPosition.Left, if it's undefined then it won't be equal anyways

ranasaria replied:
done

ranasaria commented:
> You have failing hygiene checks. I don't know how you're submitting these commits but please make sure you're not bypassing the commit hooks when pushing since those will catch these kinds of issues.

I need to figure out why these are getting skipped. I usually check-in from vs code GIT source control extension. From there these checks seem to get skipped for now. From the command line, these are still run..

alanrenmsft commented on extensions/resource-deployment/src/interfaces.ts:
```diff
@@ -100,23 +100,26 @@ export interface DeploymentProviderBase {
-	summaryPage: NotebookWizardPageInfo;
+	pages: PageInfoBase[];
+	summaryPage: PageInfoBase;
 	generateSummaryPage: boolean;
```
generateSummaryPage property is not needed, right?

alanrenmsft commented on extensions/resource-deployment/src/interfaces.ts:
```diff
@@ -100,23 +100,26 @@ export interface DeploymentProviderBase {
-	pages: NotebookWizardPageInfo[];
-	summaryPage: NotebookWizardPageInfo;
+	pages: PageInfoBase[];
+	summaryPage: PageInfoBase;
```
why can't this just be part of pages property?

ranasaria replied:
it is a wizard level property whether to generate an automatic summaryPage. I will rename it to make it clearer.

alanrenmsft commented on extensions/resource-deployment/src/interfaces.ts:
```diff
@@ -153,18 +156,32 @@ export interface DialogInfoBase {
+	[key: string]: string;
+};
+
+export interface InheritableAttributes {
```
this is the common properties for field, why not just call it FieldInfoBase?

alanrenmsft commented on extensions/resource-deployment/src/interfaces.ts:
```diff
@@ -153,18 +156,32 @@ export interface DialogInfoBase {
 	labelWidth?: string;
 	inputWidth?: string;
 	labelPosition?: LabelPosition; // Default value is top
+	rowWidth?: string;
```
the term row here might not be appropriate, fieldWidth?

ranasaria replied:
I have renamed to fieldWidth for this check-in. However, labelWidth, inputWidth, fieldWidth is all a lot of widths. I will factor out width, height, layout etc as separate properties that can be applied to label, field, input component, row etc. But I will do that in a separate check-in so that I can keep various aspects of functionality independent.

alanrenmsft commented on extensions/resource-deployment/src/interfaces.ts:
```diff
@@ -244,11 +261,13 @@ export enum FieldType {
 	FilePicker = 'file_picker',
-	KubeClusterContextPicker = 'kube_cluster_context_picker'
+	KubeClusterContextPicker = 'kube_cluster_context_picker',
+	LinkedText = 'linked_text'
```
I don't think we need this, you should be able to just extend the ReadonlyText.

alanrenmsft replied:
i am still not convinced, but this is not a big deal. is the name proper? should it be hyperlink_text?

ranasaria replied:
Can i keep the feature only in linked_text and remove from the readonly_text. Readonly_text is overloaded with linking functionality which makes usage harder.

ranasaria replied:
Can i keep the feature only in linked_text and remove from the readonly_text. Readonly_text is overloaded with linking functionality which makes usage harder.

alanrenmsft replied:
@ranasaria , I don't think it is necessary, to make it easier, we just need to add a sample in: samples/sample-resource-deployment

, btw, please add a sample for your new json based wizard.

ranasaria replied:
I will add a sample for thee new JSON based wizard in samples/sample-resource-deployment. After chatting offline with alan, we decided to remove the linking feature from readonly_text FieldType so that readonly_text mainly does only the thing of allowing a user to put two different TextComponents as a field.

ranasaria replied:
I will also rename 'linked_text' to 'hyperlinked_text'

Charles-Gagnon replied:
I really don't understand the need for so many text fields. Why isn't there just one text field that has properties for the different types, such as isReadOnly, links, etc...

Maybe a bit late to be changing it here - but let's try to make sure we're thinking of stuff like this in the future when we're doing work like this so we don't end up with a huge set of components that are all basically doing the same thing.

ranasaria replied:
Consolidate ReadOnlyText, EvaluatedText, and HyperlinkedText into a single FieldType of ReadOnlyText and also consolidated dropdown options and radioOptions into a single FieldType

alanrenmsft commented on extensions/resource-deployment/src/services/notebookService.ts:
```diff
@@ -54,6 +55,24 @@ export class NotebookService implements INotebookService {
+	 * @param insertionPosition - the position at which cells are inserted. Default is a new cell at the beginning of the notebook.
+	 */
+	public async insertCellAndLaunchNotebook(notebook: string, cellStatements: string[], insertionPosition: number = 0): Promise<void> {
+		const notebook_1 = await this.launchNotebook(notebook);
```
notebook_1? haven't seen this naming convention in ADS code base.

alanrenmsft commented on extensions/resource-deployment/src/services/notebookService.ts:
```diff
@@ -34,6 +34,7 @@ export interface NotebookExecutionResult {
 
 export interface INotebookService {
 	launchNotebook(notebook: string | NotebookInfo): Thenable<azdata.nb.NotebookEditor>;
+	insertCellAndLaunchNotebook(notebook: string | NotebookInfo, cellStatements: string[], insertionPosition?: number): Promise<void>;
```
you can still use launchNotebook as the method name, just add an override to it.

alanrenmsft commented on extensions/resource-deployment/src/ui/modelViewUtils.ts:
```diff
@@ -235,11 +257,26 @@ export function createSection(context: SectionContext): azdata.GroupContainer {
 
+function processRow(rowInfo: RowInfo, context: SectionContext): azdata.Component {
+	const rowItems: azdata.Component[] = [];
+	if ('fields' in rowInfo.fields[0]) { // rowInfo.fields is RowInfo[]
```
which scenario are you trying to support here? fields inside a field? at the first glance, this doesn't seem correct for me.

alanrenmsft commented on extensions/resource-deployment/src/interfaces.ts:
```diff
@@ -244,11 +261,13 @@ export enum FieldType {
 	Options = 'options',
 	RadioOptions = 'radio_options',
 	ReadonlyText = 'readonly_text',
+	SummaryText = 'summary_text',
```
what is summaryText?

alanrenmsft commented on extensions/resource-deployment/src/ui/notebookWizard/notebookWizardPage.ts:
```diff
@@ -7,43 +7,53 @@ import { EOL } from 'os';
 
 export class NotebookWizardPage extends WizardPageBase<NotebookWizard> {
 	private inputComponents: InputComponents = {};
+	private inputComponentOriginalValues!: Map<string, string>;
```
hmm, i am still not sure why you want to keep the original values? for summary page, on page enter event, you just set the text components to the latest values, am i missing something?

alanrenmsft replied:
the key of the inputComponents is the variable name, isn't that enough?

Charles-Gagnon replied:
Yeah I don't understand why you need this mapping - the input variable name shouldn't be changing...

ranasaria replied:
The thing is that the template is stored in TextComponent.Value which changes and the original template is lost when $(variableNames) gets interpolated. 

So the evaluated_text can have something like: "This is a resource in Resourgroup:$(VAR_ResourceGroup) in subscription: $(VAR_Subscription)", specified as defaultValue. This defaultValue get put into the TextComponent' Value field as it is so the TextComponent.Value is: "This is a resource in Resourgroup:$(VAR_ResourceGroup) in subscription: $(VAR_Subscription)".

Now when this page gets loaded if the VAR_ResourceGroup was set as "RG1" and VAR_Subscription was set as "S1" then we can interpolate and change the value of TextComponent.Value to "This is a resource in Resourgroup:RG1 in subscription: S1". So everything works so far. But if the user now goes back and changes VAR_ResourceGroup to "RG2" when we load the said page then we have load the original "This is a resource in Resourgroup:$(VAR_ResourceGroup) in subscription: $(VAR_Subscription)". string and we can no longer replace the RG1 with RG2 to show them the correct value. Saving the original TextComponent.value in that maps allows to do the variable value substitution as many times as needed.

ranasaria replied:
Alan and I chatted online, so what we discussed was that instead of saving just the defaultValue of that field, we are going to save the whole FieldInfo object in the InputComponents array itself which can be used for this evaluation

ranasaria commented:
The hygiene check failure is a ghost failure. The said file is already formatted.

Charles-Gagnon commented on extensions/resource-deployment/src/interfaces.ts:
```diff
@@ -153,18 +153,33 @@ export interface DialogInfoBase {
+export interface PageInfoBase extends FieldInfoBase {
 	title: string;
 	isSummaryPage?: boolean;
+	isSummaryPageAutoGenerated?: boolean;
```
Why is this a property on every page? Shouldn't this just be at the wizard config level?

ranasaria replied:
Will move.

ranasaria replied:
fixed

Charles-Gagnon commented on extensions/resource-deployment/src/ui/deployClusterWizard/deployClusterWizard.ts:
```diff
@@ -137,16 +138,10 @@ export class DeployClusterWizard extends WizardBase<DeployClusterWizard, DeployC
-		}, (error) => {
-			vscode.window.showErrorMessage(error);
+		const variableValueStatements = this.model.getCodeCellContentForNotebook();
+		const insertionPosition = 5;
```
Comment here explaining why 5.

ranasaria replied:
will do.

ranasaria replied:
Done.

Charles-Gagnon commented on extensions/resource-deployment/src/ui/model.ts:
```diff
@@ -32,6 +33,33 @@ export class Model {
+			.filter(propertyName => propertyName.startsWith(NoteBookEnvironmentVariablePrefix))
+			.map(propertyName => {
+				const value = this.escapeForNotebookCodeCell(this.getStringValue(propertyName, ''));
+				const varName = propertyName.replace(regex, '').toLocaleLowerCase();
```
This is only going to replace the first occurrence, is that intended?

ranasaria replied:
no. I will put the 'g' flag in regex.

ranasaria replied:
Yes the regex matches NoteBookEnvironmentVariablePrefix at the beginning of the string so only one replacement is expected.

Charles-Gagnon commented on extensions/resource-deployment/src/ui/notebookWizard/notebookWizardPage.ts:
```diff
@@ -7,43 +7,68 @@ import { EOL } from 'os';
 
 export class NotebookWizardPage extends WizardPageBase<NotebookWizard> {
 	private inputComponents: InputComponents = {};
+	private inputComponentOriginalValues!: Map<string, string>;
```
Also you're marking this as non-null but in the constructor you're only setting it in a certain case. So the non-null status is just wrong here. You shouldn't use that unless it's truly actually known to be safe to use without a null check - in this case I would suggest just using the null-coalescing operation instead when you call this below to get rid of the linting errors.

ranasaria replied:
removed inputComponentOriginalValues!

ranasaria commented on extensions/resource-deployment/src/ui/deployClusterWizard/pages/summaryPage.ts:
```diff
@@ -48,13 +48,12 @@ export class SummaryPage extends WizardPageBase<DeployClusterWizard> {
 							label: localize('deployCluster.Kubeconfig', "Kube config"),
 							defaultValue: this.wizard.model.getStringValue(VariableNames.KubeConfigPath_VariableName),
-							labelFontWeight: FontWeight.Bold
+							labelCSSStyles: { labelFontWeight: FontWeight.Bold }
```
I had already spotted and fixed this.

ranasaria commented on extensions/resource-deployment/src/ui/model.ts:
```diff
@@ -32,6 +33,29 @@ export class Model {
+	 * The statements returned are escaped for use in cell of a python notebook.
+	 */
+	public getCodeCellContentForNotebook(): string[] {
+		const regex = new RegExp(`^${NoteBookEnvironmentVariablePrefix}`);
```
ok will do.

ranasaria replied:
fixed.

chlafreniere commented on extensions/resource-deployment/src/services/notebookService.ts:
```diff
@@ -59,9 +77,9 @@ export class NotebookService implements INotebookService {
 	 */
-	launchNotebookWithContent(title: string, content: string): Thenable<azdata.nb.NotebookEditor> {
+	async launchNotebookWithContent(title: string, content: string): Promise<azdata.nb.NotebookEditor> {
 		const uri: vscode.Uri = vscode.Uri.parse(`untitled:${title}`);
```
This isn't awaited properly it looks like.

ranasaria replied:
fixed

chlafreniere commented on extensions/resource-deployment/src/services/notebookService.ts:
```diff
@@ -48,9 +49,26 @@ export class NotebookService implements INotebookService {
+	 * @param cellStatements - array of statements to be inserted in a cell
+	 * @param insertionPosition - the position at which cells are inserted. Default is a new cell at the beginning of the notebook.
+	 */
+	async launchNotebookWithEdits(notebook: string, cellStatements: string[], insertionPosition: number = 0): Promise<void> {
```
This is called without being awaited in a sync method.

ranasaria replied:
fixed

Charles-Gagnon commented on extensions/machine-learning/package.json:
```diff
@@ -1,153 +1,158 @@
-  },
-  "activationEvents": [
-    "onCommand:ml.command.managePackages",
+	"name": "machine-learning",
```
Undo these changes - this shouldn't be a part of your PR

Charles-Gagnon commented on extensions/resource-deployment/src/ui/deployClusterWizard/pages/azureSettingsPage.ts:
```diff
@@ -38,9 +38,7 @@ export class AzureSettingsPage extends WizardPageBase<DeployClusterWizard> {
-					label: '',
-					labelWidth: '0px',
-					defaultValue: localize('deployCluster.SubscriptionHelpText', "{0}"),
+					label: localize('deployCluster.SubscriptionHelpText', "{0}"),
```
The localize call here is unnecessary if it's just going to be a placeholder value

ranasaria replied:
It was prexising but will remove

ranasaria replied:
fixed.

Charles-Gagnon commented on extensions/resource-deployment/src/ui/deployClusterWizard/pages/azureSettingsPage.ts:
```diff
@@ -79,9 +77,7 @@ export class AzureSettingsPage extends WizardPageBase<DeployClusterWizard> {
-					label: '',
-					labelWidth: '0px',
-					defaultValue: localize('deployCluster.LocationHelpText', "{0}"),
+					label: localize('deployCluster.LocationHelpText', "{0}"),
```
localize is unnecessary

ranasaria replied:
fixed

Charles-Gagnon commented on extensions/resource-deployment/src/ui/modelViewUtils.ts:
```diff
@@ -506,18 +627,41 @@ function processFilePickerField(context: FieldContext, defaultFilePath?: string)
 	return { input: input, browseButton: browseFileButton };
 }
 
+function getClusterContexts(file: string) {
```
return type

Charles-Gagnon commented on extensions/resource-deployment/src/utils.ts:
```diff
@@ -30,12 +31,18 @@ export function setEnvironmentVariablesForInstallPaths(tools: ITool[]): void {
+
+export function assert(condition: any, message?: string): asserts condition {
+	if (!condition) {
+		throw new AssertionError({ message: message });
```
Just throw a normal Error, assert is more of a testing library. 

Also the few calls to this that I saw didn't pass in a message which makes the error much less useful.

ranasaria replied:
fixed.

Charles-Gagnon commented on extensions/resource-deployment/src/utils.ts:
```diff
@@ -30,12 +31,18 @@ export function setEnvironmentVariablesForInstallPaths(tools: ITool[]): void {
+	}
+}
+
+export function assert(condition: any, message?: string): asserts condition {
```
Why not have condition be a boolean?

ranasaria replied:
It was to allow a truthy expression. Will change.

ranasaria replied:
fixed.

Charles-Gagnon commented on samples/sample-resource-deployment/package.nls.json:
```diff
@@ -1,5 +1,5 @@
 {
-	"extension-displayName": "Sample resource deployment extension for Azure Data Studio",
+	"extension-displayName": "Sample resource deployment extension for  Data Studio",
```
Why did you delete Azure?

ranasaria replied:
fixed.

ranasaria replied:
Adding back. It was inadvertent. Adding back.

alanrenmsft commented on extensions/resource-deployment/src/ui/deployClusterWizard/deployClusterWizard.ts:
```diff
@@ -135,27 +144,24 @@ export class DeployClusterWizard extends WizardBase<DeployClusterWizard, DeployC
 		if (this.model.authenticationMode === AuthenticationMode.ActiveDirectory) {
 			env[VariableNames.DomainServiceAccountPassword_VariableName] = this.model.getStringValue(VariableNames.DomainServiceAccountPassword_VariableName);
 		}
+		setEnvironmentVariablesForInstallPaths(tools, env);
```
why are we passing the path using environment variables? it is only meant for passwords.

ranasaria replied:
Is there a downside to this?

ranasaria replied:
the paths are needed because tools may not have been added to the global %PATH% variable if they were installed by ads.

Charles-Gagnon replied:
Wouldn't this be a problem for all the Notebooks? I thought we told users that restarting ADS may be necessary after installing tools to pick up the updated PATH.

And couldn't you just update the PATH as well? Seems like a more standard way of handling that.

ranasaria replied:
On closer look with Alan it is not needed here at all because we are adding the necessary tools path variables directly as Python code within notebook which is avaialble both when executing from the notebook and from background execution. So this indeed superfluous here and not needed. Going to remove this.

ranasaria replied:
Removed this call.

alanrenmsft commented on extensions/resource-deployment/src/ui/deployClusterWizard/deployClusterWizardModel.ts:
```diff
@@ -166,16 +166,13 @@ export class DeployClusterWizardModel extends Model {
+		const env: NodeJS.ProcessEnv = {};
+		setEnvironmentVariablesForInstallPaths(tools, env);
+		statements.push(`os.environ["${kubeCtlEnvVarName}"] = "${this.escapeForNotebookCodeCell(env[kubeCtlEnvVarName]!)}"`);
+		statements.push(`os.environ["PATH"] = os.environ["PATH"] + "${delimiter}" + "${this.escapeForNotebookCodeCell(env[ToolsInstallPath]!)}"`);
```
this.escapeForNotebook... should not compile here, right? you have moved it out to a different class

ranasaria replied:
Will double check

alanrenmsft commented on samples/sample-resource-deployment/package.json:
```diff
@@ -10,7 +10,7 @@
   "engines": {
     "vscode": "*",
-    "azdata": ">1.10.0"
+    "azdata": ">=1.10.0"
```
this needs to be 1.19 now in order to use the wizard notebook, right?

ranasaria commented:
> could you please explain why you need to pass the tool path using environment variables? also have tested the deploy BDC scenarios to make sure the generated notebooks are the same as before?

I have tested the generated notebooks by running them.

ranasaria commented:
> could you please explain why you need to pass the tool path using environment variables? also have tested the deploy BDC scenarios to make sure the generated notebooks are the same as before?

I have addressed your feedback around tools path, can you please take a look again?

alanrenmsft commented on extensions/resource-deployment/src/ui/deployClusterWizard/pages/serviceSettingsPage.ts:
```diff
@@ -400,6 +401,7 @@ export class ServiceSettingsPage extends WizardPageBase<DeployClusterWizard> {
 
 	public onLeave(): void {
 		setModelValues(this.inputComponents, this.wizard.model);
+		Object.assign(this.wizard.inputComponents, this.inputComponents);
```
what is this line for?

ranasaria replied:
it copies the inputComponents of the page to the inputComponents of the whole wizard.

alanrenmsft commented on extensions/resource-deployment/src/utils.ts:
```diff
@@ -30,12 +30,18 @@ export function setEnvironmentVariablesForInstallPaths(tools: ITool[]): void {
+	}
+}
+
+export function assert(condition: boolean, message?: string): asserts condition {
```
what does the 'asserts condition' at the end means

ranasaria replied:
This means wherever this assert function is called it acts a type guard check corresponding to the 'condition' expression.

alanrenmsft commented on extensions/resource-deployment/src/ui/deployClusterWizard/deployClusterWizardModel.ts:
```diff
@@ -166,16 +166,13 @@ export class DeployClusterWizardModel extends Model {
-		statements.push(`os.environ["${kubeCtlEnvVarName}"] = "${this.escapeForNotebookCodeCell(process.env[kubeCtlEnvVarName]!)}"`);
-		statements.push(`os.environ["PATH"] = os.environ["PATH"] + "${delimiter}" + "${this.escapeForNotebookCodeCell(process.env[ToolsInstallPath]!)}"`);
+		const env: NodeJS.ProcessEnv = {};
+		setEnvironmentVariablesForInstallPaths(tools, env);
```
is this really required? i thought we have set the tool path when we were checking the tools before the wizard is launched.

ranasaria replied:
yes because the tools should not be setting OS.environ["PATH"] which older code requires. I thought you wanted to remove reliance on setting environment variables except for passwords. With this change in place, tools installation and discovery do not need to set process.env[ToolsInstallPath] and process.env[kubeCtlEnvVarName] which the old code depends on. 

If you would rather that I not make this change, I can leave it as it is as this is just cleaning up some usage.

ranasaria replied:
The setting of toolspath in process.env has changed with the change to in this PR:
extensions/resource-deployment/src/ui/resourceTypePickerDialog.ts

We no longer set process.env variables for anything. only wizards and dialogtabs will set process.env variables and that too only for passwords.

ranasaria replied:
So I would like to keep this change unless you insist that we continue to keep the previous code.
